File: xipc.js
const fs = require("fs");
const net = require("net");
const child_process = require("child_process");

const xipc = () => ({});
const makeClient = (port) => {
  return new Promise((resolve, reject) => {
    const client = new net.Socket();
    client.connect(port, () => {
      resolve(client);
    });
  });
};
const mod = (client) => {
  return new Proxy(
    {},
    {
      get: function (_, name) {
        return function () {
          const argz = new Array(arguments.length).fill(null);
          let i = 0;
          for (; i < argz.length; i++) {
            argz[i] = arguments[i];
          }
          const toWrite = name + ";" + JSON.stringify(argz, "utf-8");
          return new Promise((res, rej) => {
            let l = (d) => {
              res(JSON.parse(d.toString("utf-8")));
              client.removeListener("data", l);
            };
            client.on("data", l);
            client.write(Buffer.from(toWrite, "utf-8"));
          });
        };
      },
      set: function () {},
    }
  );
};

const hack = function () {
  const scripts = arguments;
  return async function (f) {
    const start_port = 8001;
    let end_port = 8001;
    let port = start_port;
    const out = xipc();
    const dr = fs.readdirSync(".");
    let i = 0;
    for (; i < scripts.length; i++) {
      const script = scripts[i];
      if (dr.indexOf(script + ".py") !== -1) {
        const fn = ".xipc." + port + ".py";
        fs.writeFileSync(
          fn,
          child_process.execSync("python py_server.py " + script + " " + port)
        );
        child_process.spawn("python", [fn], {
          detached: true,
          stdio: "ignore",
          windowsHide: true,
        });
      } else if (dr.indexOf(script + ".js") !== -1) {
        const fn = ".xipc." + port + ".js";
        fs.writeFileSync(
          fn,
          child_process.execSync("python js_server.py " + script + " " + port)
        );
        child_process.spawn("node", [fn], {
          detached: true,
          stdio: "ignore",
          windowsHide: true,
        });
      } else {
        throw Exception("Cannot handle " + script);
      }
      const client = await makeClient(port);
      out[script] = mod(client);
      port += 1;
    }
    end_port = port;
    try {
      await f(out);
    } catch (e) {
      console.log(e);
    }
    i = start_port;
    for (; i < end_port; i++) {
      const client = new net.Socket();
      client.connect(i, () => {
        client.write(";");
        client.destroy();
      });
    }
    process.exit(0);
  };
};

module.exports = { hack: hack };

File: xipc.py
import json
import logging
import os
import socket
import subprocess
from dataclasses import dataclass
from typing import Any, Sequence

logger = logging.getLogger(__name__)


class xipc:
    pass


def mod(port):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect(("127.0.0.1", port))

    class _mod:
        def __getattribute__(self, name: str) -> Any:
            def callable(*args):
                argz = json.dumps(args)
                s.sendall((name + ";" + argz).encode("utf-8"))
                return json.loads(s.recv(2048).decode("utf-8"))

            return callable

    return _mod(), s


class xipc:
    start_port = 8001
    end_port = 8001

    def __init__(self, *scripts: str):
        self._scripts = scripts
        self._sox = []

    def __enter__(self):
        logger.debug("starting xipc")
        port = self.start_port
        out = xipc()
        dr = os.listdir(".")
        self._sox = []
        for script in self._scripts:
            if ("%s.py" % script) in dr:
                fn = ".xipc.%d.py" % port
                with open(fn, "w") as fi:
                    fi.write(
                        subprocess.check_output(
                            ["python", "py_server.py", script, str(port)]
                        ).decode("utf-8")
                    )
                    subprocess.Popen(
                        ["python", fn],
                        stderr=subprocess.DEVNULL,
                        stdout=subprocess.DEVNULL,
                    )
            elif ("%s.js" % script) in dr:
                fn = ".xipc.%d.js" % port
                with open(fn, "w") as fi:
                    fi.write(
                        subprocess.check_output(
                            ["python", "js_server.py", script, str(port)]
                        ).decode("utf-8")
                    )
                    subprocess.Popen(
                        ["node", fn],
                        stderr=subprocess.DEVNULL,
                        stdout=subprocess.DEVNULL,
                    )
            else:
                raise ValueError("Cannot handle %s" % script)
            modl, sok = mod(port)
            self._sox.append(sok)
            setattr(out, script, modl)
            port += 1
        self.end_port = port
        return out

    def __exit__(self, type, value, traceback):
        for s in self._sox:
            s.sendall(";".encode("utf-8"))
            s.close()
